### ISSUE TITLE
Contour storage and recovery in rtdc file format

### ISSUE THREAD

**Issue by chrherold** [ZELLMECHANIK-DRESDEN/dclab#48]
Contour columns are not present in rtdc files originating from ShapeIn for performance reasons.

Instead, a second image set called "mask" is included. It consists of arrays with values of 0 outside the event contour and 255 within the event contour.

We should check if a transformation back to contour arrays is possible with reasonable performance in ShapeOut

The contour array can be recovered from this "binary" image using the openCV function 
https://docs.opencv.org/2.4/modules/imgproc/doc/structural_analysis_and_shape_descriptors.html

findContours 
using  mode CV_RETR_EXTERNAL
and method CV_CHAIN_APPROX_NONE

to store contours in equal fashion to rtdc files from dclab the following function can be used

drawContours

using 
image: of size of the event image filled with zeros to begin
contours: the contour array
color: Scalar(255)
thickness: FILLED

Would you make this ancillary feature that is only performed when needed? E.g. if somebody clicks an event and the image and contour should be displayed, or for every event in case columns that need this information, like volume, are used?

Here is an exemplary rtdc file as created by ShapeIn:

[M1_data.rtdc.zip](https://github.com/ZELLMECHANIK-DRESDEN/dclab/files/1927626/M1_data.rtdc.zip)

**Comment by paulmueller** [comment]
It would be easiest, if the contours were computed at once (as an ancillary column). However, I can see that this might result in a performance hit, in which case it might be better to compute contours on-the-fly. I will let you know once I have more information.

**Comment by paulmueller** [comment]
@chrherold 

I am confused as to why we would need the findContours functionality from OpenCV. 
I think a simple binary erosion would correctly retrieve the contour from the binary image, e.g.:

```python
import numpy as np
import scipy.ndimage as spim

import matplotlib.pylab as plt

x = np.linspace(-1, 1, 100).reshape(-1,1)
y = x.reshape(1,-1)

im1 = x**2 + y**2 < .5**2
im2 = (x-.5)**2 + y**2 < .2**2
im = im1 + im2

imer = spim.binary_erosion(im)

plt.subplot(121, title="mask")
plt.imshow(im)

plt.subplot(122, title="contour from mask")
plt.imshow(im-imer)

plt.show()
```

Is there anything that I am missing?

In any case, could you please give me a full dataset with "mask" and "contour" (any file format), so that I can test the transformation?

**Comment by paulmueller** [comment]
@chrherold 

I am confused as to why we would need the findContours functionality from OpenCV. 
I think a simple binary erosion would correctly retrieve the contour from the binary image, e.g.:

```python
import numpy as np
import scipy.ndimage as spim

import matplotlib.pylab as plt

x = np.linspace(-1, 1, 100).reshape(-1,1)
y = x.reshape(1,-1)

im1 = x**2 + y**2 < .5**2
im2 = (x-.5)**2 + y**2 < .2**2
im3 = (x+.6)**2 + (y-.01)**2 < .013**2
im = im1 + im2 + im3

imer = spim.binary_erosion(im)

plt.subplot(121, title="mask")
plt.imshow(im)

plt.subplot(122, title="contour from mask")
plt.imshow(im ^ imer)


plt.show()
```
![mask](https://user-images.githubusercontent.com/2853022/39331895-0c366dda-49a6-11e8-9ff6-e0dda71e6cda.jpg)

Is there anything that I am missing?

In any case, could you please give me a full dataset with "mask" and "contour" (any file format), so that I can test the transformation?

**Comment by chrherold** [comment]
@paulmueller 

We can also directly store the "mask" not as mask but as the eroded contour with 255 values at the contour only, no difference on my end.

Is this preferred?

**Comment by paulmueller** [comment]
No, let's leave it at the filled contour, then we have a clear distinction between "mask" and "contour". I just wanted to make sure that the actual task is to determine the order of the contour positions.

**Comment by paulmueller** [comment]
TODO:
- [ ] Implement computation of contour from mask
- [ ] Test hierarchy format with mask
- [ ] Basic tests for mask
- [ ] Test rtdc writer with mask
- [ ] Allow to disable export of contour data in rtdc format if only mask is given in the original data (ShapeOut export)?

**Comment by paulmueller** [comment]
I used an implementation from scikit-image to determine the contour coordinates. I would like to have your opinion regarding the error made:

These are the ancillary features that rely on contour computation. The difference compared to the OpenCV code is small for volume computation (some outliers) and for the convex area ratio. The error in the raw area ratio is higher, but no one is supposed to use it anyway, right?
![comparison_features](https://user-images.githubusercontent.com/2853022/39727535-5bdc557e-5253-11e8-9cd1-4a78a2b65844.jpg)

Here are a few images and masks of the data with largest outliers. It looks like these are unimportant events anyway, correct?
![comparison_images](https://user-images.githubusercontent.com/2853022/39727536-5bfd855a-5253-11e8-9ce3-7a5f9bab1db3.jpg)


This is the python script (only works with the latest development version):

```python
import numpy as np
import matplotlib
import matplotlib.pylab as plt
import dclab

ds1 = dclab.new_dataset("M3_data.rtdc")

feats = ds1.features
feats.remove("contour")
feats.remove("volume")
feats.remove("inert_ratio_cvx")
feats.remove("inert_ratio_raw")
ds1.export.hdf5("M3_data_nocontour.rtdc", features=feats, override=True)

ds2 = dclab.new_dataset("M3_data_nocontour.rtdc")

v1 = ds1["volume"]
v2 = ds2["volume"]

ic1 = ds1["inert_ratio_cvx"]
ic2 = ds2["inert_ratio_cvx"]

ir1 = ds1["inert_ratio_raw"]
ir2 = ds2["inert_ratio_raw"]

# plot differences
plt.figure(figsize=(6,10))

ax1 = plt.subplot(311)
plt.plot(v1, (v2-v1), "o", alpha=.3)
ax1.set_xlabel("volume")
ax1.set_ylabel("volume error")

ax2 = plt.subplot(312)
plt.plot(ic1, (ic2-ic1), "o", alpha=.3)
fmt=matplotlib.ticker.ScalarFormatter(useOffset=False)
fmt.set_scientific(False)
ax2.yaxis.set_major_formatter(fmt)
ax2.set_xlabel("inert_ratio_cvx")
ax2.set_ylabel("inert_ratio_cvx error")

ax3 = plt.subplot(313)
plt.plot(ir1, (ir2-ir1), "o", alpha=.3)
ax3.set_xlabel("inert_ratio_raw")
ax3.set_ylabel("inert_ratio_raw error")

plt.tight_layout()
plt.savefig("comparison_features.jpg")
plt.close()

# plot images where deviation is largest:
fig, axes = plt.subplots(6,1, figsize=(4,10))

idv = np.argmax(v2-v1)
axes[0].imshow(ds1["image"][idv], cmap="gray")
axes[1].imshow(ds1["mask"][idv], cmap="gray")

idc = np.argmax(ic2-ic1)
axes[2].imshow(ds1["image"][idc], cmap="gray")
axes[3].imshow(ds1["mask"][idc], cmap="gray")

idr = np.argmax(ir2-ir1)
axes[4].imshow(ds1["image"][idr], cmap="gray")
axes[5].imshow(ds1["mask"][idr], cmap="gray")

plt.tight_layout()
plt.savefig("comparison_images.jpg")
```

**Comment by phidahl** [comment]
I think the erode-based algorithm is not equivalent to the border following implemented in cv2.findContours(). Erosion can lead to loss of information regarding small 1-pixel-wide features.
For proper interpretation of the saved contours I would prefer usage of the opencv algorithm. If you don't want to waste disk space, it might be possible to extract only the needed function from opencv? OpenCV uses the algorithm described a paper by Suzuki. https://www.sciencedirect.com/science/article/pii/0734189X85900167

**Comment by paulmueller** [comment]
1-pixel-wide features are not lost in binary erosion (see example images a few posts up). The problem is that the contour-finding algorithm I borrowed from scikit-image (with a few modification such that we get contour coordinates on a pixel grid) is not equivalent to cv2.findContours().

The cv2 implementation is here: https://github.com/opencv/opencv/blob/master/modules/imgproc/src/contours.cpp

If we want to use that one, I will have to put in some time to identify the part of the code that does contour extraction (either from the mask or the eroded mask) and then translate it to python.
BTW this is not just about disk space. It is about being able to test and build dclab on all platforms without having to hack around in the build environment..

**Comment by paulmueller** [comment]
@phidahl @chrherold 

I take it the deviations from cv2.findContours shown above are too large for usage in production?

**Comment by chrherold** [comment]
@paulmueller 

I think I would like to better understand what makes the difference.

The examples you showed are indeed no nice cell events and it looks to me like all of them have some "single pixel wide" bridges or extensions.

While this is surely not the normal cell event (and everything seems to look good for normal cell events), I strongly suggest not to introduce a method that would give different results even on very exotic events, because we cannot foresee what user might want to measure and analyze.
Even if an event of a single pixel line brings all kind of questions on the use and validity of its analysis it should be treated consistently by all our software.

I observe two things:
1)
Using the convex hull seems to even out the differences (looks like rounding issues only?), i.e. the convex hull contours are the same both ways.

2)
The method you are trying seems to give larger volumes and in most cases a larger extension in the flow (or x) direction. To me this means the contour found by your test algorithm is extending further than what is returned by openCV.
It would be good to find out if this is something related to the eroding step or the translation of the eroded contour in the image to data points of the contour. (It might well be possible that the openCV way is the less "exact")
Could you produce some comparison by creating a two channel image of the original mask and the eroded mask and in another instance this image also overlaid with drawing the original contour and the contour recovered by your algorithm?

Thanks
Christoph

PS: the "inertia ratio raw" is the more important one of the two.

**Comment by paulmueller** [comment]
I hope I did not open another one of pandora's boxes :sweat:.
I will come back here with a proper comparison of the contour images.

**Comment by paulmueller** [comment]
I think I figured it out. The difference in the contour data cannot be visualized from the binary contour images (they are exactly the same). 
My implementation however, did not treat 1px-wide "contour lines" the same way as OpenCV treated them. Here is one on the left of the plotted mask:
![badmask](https://user-images.githubusercontent.com/2853022/39839262-e58b50c0-53db-11e8-822d-b2cd6b3e6040.jpg)

The part of such a contour returned by OpenCV would look like this:
```
      ...
       [ 81,  35],
       [ 80,  36],
       [ 79,  36],
       [ 78,  36],
       [ 77,  36],
       [ 76,  36],
       [ 75,  36],
       [ 74,  36],
       [ 73,  36],
       [ 74,  36],
       [ 75,  36],
       [ 76,  36],
       [ 77,  36],
       [ 78,  36],
       [ 79,  36],
       [ 80,  36],
       [ 81,  36],
       [ 82,  36],
       [ 83,  37],
       ...
```

whereas this part of the contour with my implementation would look like this:

```
       ...
       [ 81,  35],
       [ 80,  36],
       [ 79,  36],
       [ 78,  36],
       [ 77,  36],
       [ 76,  36],
       [ 75,  36],
       [ 74,  36],
       [ 73,  36],
       [ 81,  36],
       [ 82,  36],
       [ 83,  37],
       ...
``` 

In other words, the contour in OpenCV goes back-and forth along the 1px-wide "contour line". My implementation did not contain the same value twice.

I fixed that. Here are the new error plots:
![comparison_features](https://user-images.githubusercontent.com/2853022/39839135-821cec7e-53db-11e8-8b8b-b515d08fe2b5.jpg)

The outlier in the volume computation appears to be related to the order in that the contour values come. However, the error is still below 1%.

The other errors are negligible as well. I guess we can consider this resolved.

The take-home message is, that these 1-px-wide contour protuberances are bad in general and we likely are not treating them correctly.

**Comment by chrherold** [comment]
Thanks hunting down the problem. I agree that one pixel wide features are bad and lead to a ton of problems but since they are not at all what all this is made for, the most important thing is to
handle it consistently. As to what would be "correct" for a single pixel wide feature... I think this is not possible to pin down as it is simply below the resolution limit of the method.

So agreed, this can be considered solved. Thanks!
